rfcs/input-validations.md

```diff
@@ -240,67 +377,204 @@ mutation update_author {
+
+{
+    "message": "Validation Failed",
+    "details": {
+        "error": "Phone number is invalid"
+    }
+}
 ```
+Note, The `message` field is always added by Hasura in case of failed validation.
+
+When an unexpected response format is received, Hasura raises internal exception.
+
+
+### Examples
+
+1. [Single Update Condition] Check if `name` is a valid string
 
-### Delete mutation
 
 Consider the following sample mutation query:
+```graphql
+mutation update_author {
+  update_author(where: { id: { _eq: 3 } }, _set: { name: "Jane" }) {
+    affected_rows
+  }
+}
+```
+
+The arguments used to update `author` model i.e. `where` and `_set` are sent to the
+`validate_input` url of `author` update permission. The url can check the `email`
+value. If the url returns success, then the mutation proceeds else the error from the
+url is forwarded.
+
+The sample payload the validation URL of `author` model receives is:
+```json
+{
+    "version": 1,
+    "role": "user",
+    "session_variables": {
+        "x-hasura-role": "user"
+    },
+    "data": {
+      "updates": [
+          {
+              "where": {"id": {"_eq": 3}},
+              "_set": {"name": "Jane"}
+          }
+      ]
+    }
+}
+```
+
+2. [Multiple Update Condition] Check if the number of update conditions are less than 50
+
+Consider the following sample mutation query:
+```graphql
+mutation update_many_articles {
+  update_article_many (
+    updates: [
+      {
+        where: {rating: {_lte: 1}},
+        _set: {is_published: false}
+      },
+      {
+       where: {rating: {_gte: 4}},
+        _set: {is_published: true}
+      }
+    ]
+  ) {
+    affected_rows
+  }
+}
+```
+The arguments used to update `articles` model i.e. the list of update arguments
+(`where` and `_set`) are sent to the `validate_input` url of `author` update
+permission. The url can check the number of update arguments. If the url returns
+success, then the mutation proceeds else the error from the url is forwarded.
+
+The sample payload the validation URL of `article` model receives is:
+```json
+{
+    "version": 1,
+    "role": "user",
+    "session_variables": {
+        "x-hasura-role": "user"
+    },
+    "data": {
+        "updates": [
+            {
+              "where": {"rating": {"_lte": 1}},
+              "_set": {"is_published": false}
+            },
+            {
+              "where": {"rating": {"_gte": 4}},
+              "_set": {"is_published": true}
+            }
+        ]
+    }
+}
+```
+
+3. [Update condition by Primary Key] Check if `name` is a valid string
+
+Consider the following sample mutation query:
+```graphql
+mutation update_author {
+  update_author_by_pk(pk_columns: {id: 3}, _set: { name: "Jane" }) {
+    affected_rows
+  }
+}
+```
+
+The sample payload the validation URL of `author` model receives is:
+```json
+{
+    "version": 1,
+    "role": "user",
+    "session_variables": {
+        "x-hasura-role": "user"
+    },
+    "data": {
+      "updates": [
```

Comment:
Why do we call this `updates` whereas in insert spec we call it `objects`? Can we unify here?

Why do we need `objects`, `updates` or `deletes` anyways? Why not just add them after `data`?

Reply:
>  Why not just add them after data?

The reason we don't add them after `data` was to give us the oppurtunity to extend the data field to hold much more data. The `objects` field would signify the actual arguments that are given to mutation. But say in future, we want to inlcude more fields into the `data` then it'll be easier to add. Else all of the other new field would need to be added as root fields. 

> Why do we need `objects`, `updates` or `deletes` anyways?

My reasoning was purely from the point of user experience. Say a user would like to have only one webhook URL for a graphql model. Where the logic is different for different type of mutation.

For eg:   Say we have a `article` table and if we want to have all types of validation on this, the user would then create the following validation URL's in `insert/delete/update` permission:

1. Insert: http://validate.com/article/insert
2. Update: http://validate.com/article/update
3. Delete: http://validate.com/article/delete

But, that might be too hard to manage no, having ways to different between the different payload can help users.

But thinking about it now, This just seems too much overengineering  :see_no_evil:

A better option would be to: 

1. Have a `mutation_type` field in the response which will tell the type of mutation
2. Have only the `data.objects`  field for all types of mutation.

Reply:
Why not `data.input` (So it matches the feature -> input validation)? We can add `mutation_type` to the payload later?

Reply:
Thanks! have made the changes.